Allow toggling fieldset states with keyboard actions (A11Y)

zoglo opened the pull request:
This PR fixes accessibility with keyboards not being able to focus the fieldset legend in the backend when using `tabulator` and trying to open it with `enter`

#### Previous behavior

https://github.com/contao/contao/assets/55794780/10582570-8136-46a1-a05a-47ec2a15839d


#### This PR

https://github.com/user-attachments/assets/f65d5a7d-9fc4-4fb3-994a-d72b45b09034


#### More information on this PR:

- [multiple actions with stimulus](https://www.dwightwatson.com/posts/using-multiple-actions-on-a-single-element-with-stimulus)
- [using window focus](https://stimulus-use.github.io/stimulus-use/#/use-window-focus)
- [keyboard actions](https://github.com/hotwired/stimulus/issues/129#issuecomment-375314877)

zoglo committed:
Allow toggling fieldset with keyboard

zoglo committed:
Allow toggling fieldset with keyboard

zoglo commented:
Might need to apply a `tabindex="0"` (via JS) on the legacy layer as well, wdyt @aschempp ?

aschempp commented on core-bundle/assets/controllers/toggle-fieldset-controller.js:
```diff
@@ -16,7 +16,7 @@ export default class extends Controller {
             fs.setAttribute(`data-${identifier}-table-value`, table);
             fs.setAttribute(`data-${identifier}-collapsed-class`, 'collapsed');
-            el.setAttribute('data-action', `click->${identifier}#toggle`);
+            el.setAttribute('data-action', `click->${identifier}#toggle keydown->${identifier}#toggleWithKeyboard`);
```
shouldn't this work something like that:

```suggestion
            el.setAttribute('data-action', `click->${identifier}#toggle keydown.enter->${identifier}#toggle`);
```

aschempp commented:
On a side note, I'm not sure I would toggle that legend with the _enter_ key. How do you trigger a focussed button element? I  would usually use the _space_ key for that?

zoglo commented:
> On a side note, I'm not sure I would toggle that legend with the _enter_ key. How do you trigger a focussed button element? I would usually use the _space_ key for that?

https://adrianroselli.com/2022/04/brief-note-on-buttons-enter-and-space.html

Enter triggers the button, space would only work once you release - wouldn't that also mean that it'd be better to use native buttons inside the legends then?

zoglo committed:
Use keyboardEvent filter

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Use keyboardEvent filter

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Use keyboardEvent filters and allow `space` as well

zoglo committed:
Use keyboardEvent filters and allow `space` as well

zoglo committed:
Add `tabindex` to legacy layer

zoglo committed:
Add `tabindex` to legacy layer

zoglo commented:
@aschempp as mentioned above, holding down the spacebar would not trigger the action but it doesn't really make sense when toggling the fieldset to hold down spacebar and on release opening it.

- Enter clicks a button, Space clicks a button
- Holding Enter fires the button continuosly
- Holding Space only fires on release of the button

I also implemented your suggestion and additionally added a `:prevent` for the the space-keydown to prevent the default "scrolling" behavior of it.

ausi commented:
> wouldn't that also mean that it'd be better to use native buttons inside the legends then?

IMO yes. Changing `<legend data-action="…">…</legend>` 
to `<legend data-action="…"><button type="button">…</button></legend>`
should already handle all the accessibility features needed without any extra JavaScript. It would also be way more reliable I think.

zoglo commented:
> IMO yes. Changing `<legend data-action="…">…</legend>`
> to `<legend data-action="…"><button type="button">…</button></legend>`
> should already handle all the accessibility features needed without any extra JavaScript. It would also be way more reliable I think.

It would still need to collapse / open the fieldset, hence still needing the data-action and would need restyling + maybe even work for the full width of the legend and not just the link. I'd prefer that over my current solution

ausi commented:
> It would still need to collapse / open the fieldset, hence still needing the data-action

The `data-action="…"` of the `<legend>` should be enough for that. 
But I agree, styling it correctly and moving the `data-action` to the `<button>` would be the cleaner implementation.

aschempp commented on core-bundle/assets/controllers/toggle-fieldset-controller.js:
```diff
@@ -16,7 +16,8 @@ export default class extends Controller {
             fs.setAttribute(`data-${identifier}-collapsed-class`, 'collapsed');
-            el.setAttribute('data-action', `click->${identifier}#toggle`);
+            el.setAttribute('tabindex', 0);
+            el.setAttribute('data-action', `click->${identifier}#toggle keydown.enter->${identifier}#toggle keydown.space->${identifier}#toggle:prevent`);
```
```suggestion
            el.setAttribute('data-action', `click->${identifier}#toggle keydown.enter->${identifier}#toggle keyup.space->${identifier}#toggle:prevent`);
```

aschempp commented on core-bundle/contao/drivers/DC_File.php:
```diff
@@ -164,7 +164,7 @@ public function edit()
 					list($key, $cls) = explode(':', $legends[$k]) + array(null, null);
 
-					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
+					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle keydown.enter->contao--toggle-fieldset#toggle keydown.space->contao--toggle-fieldset#toggle:prevent" tabindex="0">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
```
```suggestion
					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle keydown.enter->contao--toggle-fieldset#toggle keyup.space->contao--toggle-fieldset#toggle:prevent" tabindex="0">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
```

aschempp commented on core-bundle/contao/drivers/DC_Table.php:
```diff
@@ -2039,7 +2039,7 @@ public function edit($intId=null, $ajaxId=null)
 					list($key, $cls) = explode(':', $legends[$k]) + array(null, null);
 
-					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
+					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle keydown.enter->contao--toggle-fieldset#toggle keydown.space->contao--toggle-fieldset#toggle:prevent" tabindex="0">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
```
```suggestion
					$legend = "\n" . '<legend data-action="click->contao--toggle-fieldset#toggle keydown.enter->contao--toggle-fieldset#toggle keyup.space->contao--toggle-fieldset#toggle:prevent" tabindex="0">' . ($GLOBALS['TL_LANG'][$this->strTable][$key] ?? $key) . '</legend>';
```

aschempp reviewed (changes requested):
According to https://adrianroselli.com/2022/04/brief-note-on-buttons-enter-and-space.html, the space bar should be handled on keyup. 

Otherwise I'd agree a button makes more sense in terms of accessibility. Not sure how easy that would be to add, e.g. also for the BC layer.

zoglo committed:
Use `keyup` for spacebar

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Use `keyup` for spacebar

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Use `keyup` for spacebar event (followup to 2ea18751f44620e049eff5ef14cc4c251bcd9183)

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Use `keyup` for spacebar event (followup to 2ea18751f44620e049eff5ef14cc4c251bcd9183)

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Consider keydown prevention for `space`

zoglo committed:
Consider keydown prevention for `space`

zoglo committed:
Use native `button` for focusing legends

- also consider migrateLegacy fallback

zoglo committed:
Use native `button` for focusing legends

- also consider migrateLegacy fallback

zoglo committed:
Rebuild the assets

zoglo committed:
Rebuild the assets

zoglo commented:
@aschempp the `keyup` makes sense but the `:prevent` flag won't work on keydown anymore - I addressed it in these commits:
~~- 2ea18751f44620e049eff5ef14cc4c251bcd9183~~
~~- cdc3cf59523f3aaa522257b3510e50a20255ece3~~
~~- 1f5c7e16d65acb5313bc263e2aadcbbd2418e573~~

### Update 
I've changed it to a button as suggested in https://github.com/contao/contao/pull/7235#issuecomment-2129328690  and now it should work native.

> Otherwise I'd agree a button makes more sense in terms of accessibility. Not sure how easy that would be to add, e.g. also for the BC layer.

I've added that by creating a button and replacing the inner content as seen here:
https://github.com/zoglo/contao/blob/5ef32a490adf5831ee6aa432fdfa7ce0dec4ca58/core-bundle/assets/controllers/toggle-fieldset-controller.js#L19-L28

We may also leverage "aria-expanded" now but I am unsure if that is necessary

zoglo commented:
> But I agree, styling it correctly and moving the `data-action` to the `<button>` would be the cleaner implementation.

@ausi could you also check if this is fine for you?

Using `all: unset` makes sure the button wouldn't have any styles, however I have to readd the native `:focus-visible` styles to make sure that the focusable-attribute is being seen:

https://github.com/contao/contao/pull/7235/commits/a09fc0d595259632cdcfe26319ab5c61fca1fa3d#diff-67f03b1497b4eb95327a19cd9cf64145918242e6cd12c22b561ae3533a6584beR408-R416

aschempp reviewed (commented):
Looks good in general. I would probably not dynamically generate the button, because we don't know where people use this. Not having perfect accessibility on a BC layer is preferrable to me than changing a BC layer markup.

ausi commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -405,6 +405,15 @@ table.with-zebra tbody tr:nth-child(even) td {
+
+button[data-action="contao--toggle-fieldset#toggle"]:focus-visible {
+	outline: -webkit-focus-ring-color auto 1px;
+}
```
I think instead of adding this here, we should change the `fieldset.tl_tbox > legend, fieldset.tl_box > legend` styles in the main.css to `fieldset.tl_tbox > legend > button, fieldset.tl_box > legend > button`.
With your current solution a click on the arrow-symbol would probably not trigger the action, right?

ausi replied:
And I think `-webkit-focus-ring-color` is not supported everywhere. Maybe we need a fallback for that?

zoglo replied:
> I think instead of adding this here, we should change the `fieldset.tl_tbox > legend, fieldset.tl_box > legend` styles in the main.css to `fieldset.tl_tbox > legend > button, fieldset.tl_box > legend > button`. With your current solution a click on the arrow-symbol would probably not trigger the action, right?

Good catch, might want to restyle the whole button to full-width then with the `:has`-selector due to aschempp's comment in https://github.com/contao/contao/pull/7235#pullrequestreview-2080925161.

> And I think `-webkit-focus-ring-color` is not supported everywhere. Maybe we need a fallback for that?

See my comment above, I'd rather use proper styles now than resetting since simulating any browser-native styles across all browsers is kinda a PITA, especially with webkit / non-webkit / safari

leofeyer replied:
You do realize that we don‘t have to use buttons at all costs, right? We can use `aria-role="button"` and `aria-pressed` on non-button elements to make them accessible.

<img width="775" alt="" src="https://github.com/contao/contao/assets/1192057/15ce6330-c907-4ab1-bc13-960d69c19652">

zoglo replied:
That would still not make it focusable @leofeyer.

If not going for a native `<button>`, you'd still have to use the following markup to achieve the same result and to make it accessible which is what could be avoided by going for the native HTML solution.

**Not using a native button (Also adding the `aria-pressed` state via Stimulus)**
```html
<legend tabindex="0" aria-role="button" aria-pressed="false" data-action="click->contao--toggle-fieldset#toggle keydown.enter->contao--toggle-fieldset#toggle keyup.space->contao--toggle-fieldset#toggle:prevent">My legend</legend>
```

**Native button**
```html
<legend><button type="button" data-action="contao--toggle-fieldset#toggle">My legend</button></legend>
```

ausi replied:
AFAIK the best practice accessibility-wise is to use a `<button>` for such things instead of implementing all the accessibility features yourself (focus, keyboard navigation, ARIA role,…)

But I agree we should add `aria-pressed="…"` to the button to improve the accessibility even more.

zoglo replied:
> We may also leverage "aria-expanded" now but I am unsure if that is necessary

Probably the same here if we are talking about screen readers.

leofeyer replied:
> But I agree we should add `aria-pressed="…"` to the button to improve the accessibility even more.

That would be redundant, wouldn‘t it?

zoglo replied:
Since it indicates that the `<button>` becomes a "toggle" button, it would further enhance it:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed

It's not redundant since buttons aren't just for toggle states.

zoglo replied:
I restyled it within the last commits  a3c847d31dd894a69a76b940a05c7bcb02e9ad08 29d597edda73194519a9dff7c8f1b1aed7b16c0b

We still have to add the following states within the stimulus controller:

- aria-pressed="`true|false`" -> on the `<button>`
- aria-expanded="`true|false`" -> on the `<fieldset>`

zoglo committed:
Merge remote-tracking branch 'zoglo/bugfix/toggle-fieldset-a11y' into bugfix/toggle-fieldset-a11y

zoglo committed:
Remove BC layer button

zoglo committed:
Restyle legend to be more accessible

zoglo committed:
Merge branch '5.3' into bugfix/toggle-fieldset-a11y

zoglo commented:
Still need to address this before this is ready to merge: https://github.com/contao/contao/pull/7235#discussion_r1619539602

aschempp commented:
@zoglo anything we can do to finish this? 🙃

zoglo commented:
@aschempp 
I wrote it down here:
https://github.com/contao/contao/pull/7235#discussion_r1619539602

zoglo committed:
Merge branch '5.3' into bugfix/toggle-fieldset-a11y

zoglo committed:
Revert changes

zoglo committed:
Rebuild the assets

zoglo committed:
Handle `ariaExpanded` for fieldsets

zoglo commented:
I did not add `aria-pressed` on buttons since that one is handled by the browser itself, however there's the `aria-expanded`-attribute that I've added to make sure that screenreaders like NVDA will know after a toggle has been pressed that it is open or not.

Should be ready to review right now, please hit me up if you want to test it with NVDA or another screenreader installed and you want an online test environment with this PR :)

aschempp commented on core-bundle/assets/controllers/toggle-fieldset-controller.js:
```diff
@@ -59,20 +60,24 @@ export default class extends Controller {
     connect () {
         if (this.element.querySelectorAll('label.error, label.mandatory').length) {
             this.element.classList.remove(this.collapsedClass);
+            this.element.querySelector('button').ariaExpanded = 'false';
```
what if the button does not exist?

zoglo replied:
Good catch, we could use something like this as well, wdyt?
```js
this.element.querySelector('button')?.ariaExpanded = 'false';
```

zoglo replied:
@aschempp this is fixed in e882f0405c70dc2a62b4b3a81a38ad77b944d83d
You made me realize there was one case that wasn't added properly.

```js
    setAriaAttributes(state) {
        const button = this.element.querySelector('button');

        if (button) {
            button.ariaExpanded = state;
        }
    }
```

zoglo committed:
Only set aria attributes on existing elements

zoglo commented:
This is the state now (using NVDA for reading the attributes)

https://github.com/user-attachments/assets/f65d5a7d-9fc4-4fb3-994a-d72b45b09034

aschempp commented on core-bundle/assets/controllers/toggle-fieldset-controller.js:
```diff
@@ -16,7 +16,8 @@ export default class extends Controller {
             fs.setAttribute(`data-${identifier}-collapsed-class`, 'collapsed');
-            el.setAttribute('data-action', `click->${identifier}#toggle`);
+            el.setAttribute('tabindex', 0);
+            el.setAttribute('data-action', `click->${identifier}#toggle keydown.enter->${identifier}#toggle keydown.space->contao--toggle-fieldset#prevent:prevent keyup.space->${identifier}#toggle:prevent`);
```
```suggestion
            el.setAttribute('data-action', `click->${identifier}#toggle keydown.enter->${identifier}#toggle keydown.space->${identifier}#prevent:prevent keyup.space->${identifier}#toggle:prevent`);
```

aschempp commented on core-bundle/assets/controllers/toggle-fieldset-controller.js:
```diff
@@ -129,4 +138,12 @@ export default class extends Controller {
         });
     }
+
+    setAriaAttributes(state) {
```
```suggestion
    setAriaExpanded(state) {
```

(would need adjusting the callees as well.)

aschempp reviewed (commented):
Just some minor nitpicks 😇

zoglo committed:
Use identifier and rename method

Co-authored-by: Andreas Schempp <andreas.schempp@terminal42.ch>

zoglo committed:
Rename method from applied code review

zoglo committed:
Rebuild the assets

zoglo committed:
Merge branch '5.3' into bugfix/toggle-fieldset-a11y

aschempp reviewed (dismissed):
🥳

leofeyer committed:
Merge branch '5.3' into bugfix/toggle-fieldset-a11y

leofeyer commented on core-bundle/contao/themes/flexible/styles/main.css:
```diff
@@ -2752,7 +2752,13 @@ label.error, legend.error, .tl_checkbox_container.error legend {
 /* Fieldsets */
+fieldset legend {
+	padding-inline-start: 0;
+	padding-inline-end: 0;
```
What does that do? Unsetting these rules does not make a difference on my system.

zoglo replied:
Resetting the default behavior of the browser itself:
![image](https://github.com/user-attachments/assets/12ead102-2009-4e04-aea7-5b0c24e2fad6)

I think it's overwritten by the following lines and could be removed
https://github.com/contao/contao/blob/52fa48dd74dbf2ea528f6d0471c6f43cc3155ecd/core-bundle/contao/themes/flexible/styles/main.css#L2768-L2774

zoglo replied:
@leofeyer iirc, this must have been before I used a full button but I'm unsure. Removing it did not change anything.

leofeyer replied:
Then let’s remove it again and see what happens 😄

zoglo replied:
Already removed in https://github.com/contao/contao/pull/7235/commits/1a803e37036325ea876ff4c8e727b6f8a37f55d5 like 3hrs ago ;)

leofeyer committed:
Revert some unrelated changes

zoglo committed:
Remove obsolete css

leofeyer commented:
Thank you @zoglo.
